cli/connector/eval_plain_text.go

```diff
@@ -0,0 +1,413 @@
+	var err error
+	var evalResultEncHex string
+
+	var getResultEncHexFunc func([]byte) (string, error)
+	// Result data is returned as a table
+	// `{ data_enc = msgpack.encode(ret):hex() }`.
+	// It can't be returned as a string because of Lua output -
+	// error is showed as a string, so the returned table allows to
+	// distinguish returned string data from the error.
+	//
+	// YAML output:
+	// tarantool> return 'XXX'
+	// ---
+	// - 'XXX'
+	// ...
+	//
+	// tarantool> error('XXX')
+	// ---
+	// - error: XXX
+	// ...
+	//
+	// Lua output:
+	// tarantool> return 'XXX'
+	// "XXX";
+	// tarantool> error('XXX')
+	// "XXX";
+
+	resString := string(resBytes)
+	if strings.HasPrefix(resString, startOfYamlOutput) {
+		getResultEncHexFunc = getPlainTextEvalResYaml
+	} else {
+		getResultEncHexFunc = getPlainTextEvalResLua
+	}
+
+	if evalResultEncHex, err = getResultEncHexFunc(resBytes); err != nil {
+		return nil, err
+	}
+
+	dataEnc, err := hex.DecodeString(evalResultEncHex)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to decode hex value: %s", err)
+	}
+
+	if result != nil {
+		if err := msgpack.Unmarshal(dataEnc, result); err != nil {
+			return nil, fmt.Errorf("Failed to parse eval result: %s", err)
+		}
+
+		return nil, nil
+	}
+
+	var data []interface{}
+	if err := msgpack.Unmarshal(dataEnc, &data); err != nil {
+		return nil, fmt.Errorf("Failed to parse eval result: %s", err)
+	}
+
+	return data, nil
+}
+
+func getPlainTextEvalResYaml(resBytes []byte) (string, error) {
+	evalResults := []PlainTextEvalRes{}
+	if err := yaml.UnmarshalStrict(resBytes, &evalResults); err != nil {
+		errorStrings := make([]map[string]string, 0)
+		if err := yaml.UnmarshalStrict(resBytes, &errorStrings); err == nil {
+			if len(errorStrings) > 0 {
+				err, found := errorStrings[0]["error"]
+				if found {
+					return "", fmt.Errorf("Syntax error: %s", err)
+				}
+			}
+
+		}
+
+		return "", fmt.Errorf("Failed to parse eval result: %s", err)
+	}
+
+	if len(evalResults) != 1 {
+		return "", fmt.Errorf("Expected one result, found %d", len(evalResults))
+	}
+
+	evalResult := evalResults[0]
+	return evalResult.DataEncHex, nil
+}
+
+func getPlainTextEvalResLua(resBytes []byte) (string, error) {
+	L := lua.NewState()
+	defer L.Close()
+
+	doString := fmt.Sprintf(`res = %s`, resBytes)
+
+	if err := L.DoString(doString); err != nil {
+		return "", err
+	}
+
+	luaRes := L.Env.RawGetString("res")
+
+	if luaRes.Type() == lua.LTString {
+		return "", fmt.Errorf("Syntax error: %s", lua.LVAsString(luaRes))
+	}
+
+	encodedDataLV := L.GetTable(luaRes, lua.LString("data_enc"))
+
+	encodedData := lua.LVAsString(encodedDataLV)
+
+	return encodedData, nil
+}
+
+var (
+	callFuncTmpl = `return {{ .FunctionName }}(...)`
+)
+
+const (
+	evalFuncTmpl = `
+local function func(...)
+	{{ .FunctionBody }}
+end
+local args = require('msgpack').decode(string.fromhex('{{ .ArgsEncoded }}'))
+
+local ret = {
+	load(
+		'local func = ... return func(unpack(select(2, ...)))',
```

Comment:
Why do you need unpack here?

First of all, you could lose some values:
```
tarantool> unpack({1, nil, 2, nil, 2})
---
- 1
...
```

Secondly, you lose trailing nils.

Reply:
But...

```
tarantool> args = {1, 2, nil, 4}
---
...

tarantool> func = function(...) print(...) end
---
...

tarantool> load(
'local func = ... return func(unpack(select(2, ...)))',
'@eval'
)(func, args)
1	2	nil	4
---
...
```

Reply:
Since we use msgpack to pass arguments to instance, all nils are `box.NULL`, so there is no problems with loosing nils.